stl/inc/experimental/random

```diff
@@ -0,0 +1,50 @@
+// random experimental header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _EXPERIMENTAL_RANDOM_
+#define _EXPERIMENTAL_RANDOM_
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#include <random>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+namespace experimental {
+    inline namespace fundamentals_v3 {
+
+        inline default_random_engine& _Engine() {
+            thread_local default_random_engine _Eng{random_device{}()};

```

Comment:
I don't think we can use `thread_local` in standard headers. I know the spec talks about this, but IMO this is one of the things we need to fix in the spec before it would be ready to actually merge.
I think we should return `rand_s` or `_Random_device()` (which calls `rand_s`) instead of replacing the platform's good and secure RNG with a poor linear congruential engine...

Reply:
* What's wrong with `thread_local`, if it's mandated by the TS?
* `rand_s()` is not available within STL headers (as the CRT guards its declaration with a macro). `_Random_device()` is available, but I see no reason to avoid `random_device` as it's stateless and its default constructor no longer constructs a `string`.
* `default_random_engine` is `mt19937` for MSVC's STL:
https://github.com/microsoft/STL/blob/8f9431931b1e1f02d24d518452ae447a843c6121/stl/inc/random#L4938
* We cannot replace the TS's use of `default_random_engine` with directly using `random_device`, as that breaks the functionality of `reseed(value)`.

Reply:
>What's wrong with thread_local, if it's mandated by the TS?

The TS doesn't know that that causes a .TLS section to be emitted and that there are a limited number of those. If any .obj in kernel32.dll included the header, for example, that would critically break any program near the ~100 or so TLS slot limit.

We put other facilities that require thread-local like behavior in the vcruntime PTD structure, in part to avoid such mess.

mersenne_twister/default_random_engine are particularly problematic because they are huge, so a program that creates many threads which was just under hardware/implementation limits before might be over those limits after. In that sense, thread_local breaks the "don't pay for what you don't use" principle because most threads aren't going to use all thread_local data members.

>rand_s() is not available within STL headers

Right, I'm referring to the facility (that being Windows default cryptographically secure random engine)

>We cannot replace the TS's use of default_random_engine with directly using random_device, as that breaks the functionality of reseed(value).

Yeah, I wasn't sure if the reseed function was an artifact of the recommended implementation or a hard requirement driven by design goals. I asked on L[E]WG mailing lists and it was indicated there that the ability to reseed is indeed a design goal.

We should decide whether a created thread is supposed to derive its random state from the calling thread (like locale does) or if it should get a brand new state.